docs/plugins.md

```diff
@@ -1,5 +1,202 @@
+received from the underlying client, this execution chain is revert and the last plugin defined will be called first.
+
+i.e. with the following code: 
+
+```php
+use Http\Discovery\HttpClientDiscovery;
+use Http\Client\Plugin\PluginClient;
+use Http\Client\Plugin\RetryPlugin;
+use Http\Client\Plugin\RedirectPlugin;
+
+$retryPlugin = new RetryPlugin();
+$redirectPlugin = new RedirectPlugin();
+
+$pluginClient = new PluginClient(HttpClientDiscovery::find(), [
+    $retryPlugin,
+    $redirectPlugin
+]);
+```
+
+The execution chain will look like this:
+
+```
+Request  ---> PluginClient ---> RetryPlugin ---> RedirectPlugin ---> HttpClient ----
+                                                                                   | (processing call)
+Response <--- PluginClient <--- RetryPlugin <--- RedirectPlugin <--- HttpClient <---
+```
+
+In order to have correct behavior over the global process, you need to understand well each plugin used,
+and manage a correct order when passing the array to the `PluginClient`
+
+`RetryPlugin` will be best at the end to optimize the retry process, but it can also be good 
+to have it as the first plugin, if one in the chain is inconsistent and may need a retry.
+
+Our recommendation, which apply to most of use cases, is to organize your plugins with following rules:
+
+ 1. Plugins that modify the request should be at the beginning (like the `AuthenticationPlugin` or the `CookiePlugin`)
+ 2. Plugins which intervene in the workflow should be in the "middle" (like the `RetryPlugin` or the `RedirectPlugin`)
+ 3. Plugins which log and a debug information about a call should be last (like the `LoggerPlugin` or the `HistoryPlugin`)
+
+However there may be exception in this rule, if, for security reason, you don't want to log the authentication header, it will be better to put
+this plugin after the `LoggerPlugin` one.
+
+## Implementing your own Plugin
+
+When writing your own Plugin, you need to be aware that the `PluginClient` is async first. It means that every plugin must
+be written by respecting the `HttpAsyncClient` contract and use `Promise` as the return.
+
+Each plugin must implement the `Http\Client\Plugin\Plugin` interface.
+
+This interface defined the `handleRequest` method which allow to modify behavior of the call:
+
+```php
+/**
+ * handle the request and return the response coming from the next callable
+ *
+ * @param RequestInterface $request Request to use
+ * @param callable         $next    Callback to call to have the request, it muse have the request as it first argument
+ * @param callable         $first   First element in the plugin chain, used to to restart a request from the beginning
+ *
+ * @return Promise
+ */
+public function handleRequest(RequestInterface $request, callable $next, callable $first);
+```
+
+The `$request` is the one created by the client, you can transform it, or do what you like with it. Always be aware that
```

Comment:
we don't know who created the $request. maybe just say "The `$request` comes from upstream. You can transform it and pass a modified version downstream if you need."